Wrote script to manually import users from csv file exported from on-prem AD #64

haakondr opened the pull request:
1. exported using powershell connected to niva network on my windows machine:

```
Get-ADUser -Filter * -SearchBase "OU=OSLO,OU=NIVA,DC=niva,DC=corp" |
Select UserPrincipalName, GivenName, Surname, SamAccountName |
Export-CSV -Delimiter ";" -Encoding UTF8 -Path
"c:\niva\port\utils\data\emails.csv"
```

2. sent the file to my linux machine
3. on linux  I ran the following to fix encodings:

```
dos2unix data/emails.csv
```

4. deleted first line as it is not valid, then ran script

## features:

- duplicate detection based on person name. If person exists, we circuit
break and stop importing
- if user does not have first or last name, we ignore user (some system
users etc, we probably don't want these in the system)

## TODO:

- validate that external identifier does not already exist (as in, have
we already imported person with SamAccountName?). Currently we just
check person name, but people can change names so that is probably not
100% solid
- some fixing up, like setting connections etc based on env variables. Planning on running in nivatest when PR is approved, will fix that then on friday

haakondr commented on src/odm2_postgres_api/utils/populate_with_general_metadata.py:
```diff
@@ -21,12 +21,20 @@
+        "externalidentifiersystemurl": ""
+    }, {
+        "externalidentifiersystemname": 'onprem-active-directory',
+        "identifiersystemorganizationid": 1,
```
we are hardcoding org in a lot of places. Should probably look into fixing that

RmStorm replied:
yeah me and @ZofiaRudjord also discussed this quite a bit over the past couple days but it's not such an easy problem to solve. :disappointed:

RmStorm commented on src/import_users.py:
```diff
@@ -0,0 +1,108 @@
+            try:
+                users.append(User(**{k: name_prepare(v) for k, v in u.items()}))
+            except ValidationError:
+                pass
```
I think we should put the failed ones in a list and print at the end of the function. So we can inspect who is being ignored.

RmStorm commented on src/odm2_postgres_api/queries/core_queries.py:
```diff
@@ -7,19 +10,51 @@
+    query = f"SELECT * FROM {table} where {' AND '.join(where_clauses)}"
+    row = await conn.fetchrow(query)
+    if row:
+        return response_model(**row)
```
I think these kind of return statements look nicer as a ternary:
```python
return response_model(**row) if row else None
```

RmStorm commented on src/odm2_postgres_api/queries/core_queries.py:
```diff
@@ -7,19 +10,51 @@
     return response_model(**row)
 
 
+async def insert_pydantic_object_if_not_exists(conn: asyncpg.connection, table_name: str, pydantic_object,
```
this function and find_row really need unittests.

haakondr replied:
hm yes. I think I will not use this function so maybe I will just delete it. I don't think this was a very solid approach, could be quite error prone

RmStorm commented on src/import_users.py:
```diff
@@ -0,0 +1,108 @@
+                middlename = None
+
+            people = PeopleCreate(personfirstname=name[0], personmiddlename=middlename, personlastname=user.Surname)
+            row = await find_row(conn, "odm2.people", people.dict(), PeopleCreate)
```
let's check for the existence of emailadress in affiliation instead of people. That's probably more robust!

haakondr replied:
yes

RmStorm commented on src/import_users.py:
```diff
@@ -0,0 +1,108 @@
+
+
+async def main():
+    # db_host = os.environ["TIMESCALE_ODM2_SERVICE_HOST"]
```
maybe like so: `db_host = os.environ.get("TIMESCALE_ODM2_SERVICE_HOST", 'localhost')` for all 5

haakondr replied:
hm. I am a bit ambivalent to provide default values, but for these one-off scripts it could perhaps make sense

haakondr commented:
@RmStorm I have updated the branch with the following changes:

- automatically ignores first line of csv file (I deleted this line manually before)
- moved file into its own folder with a corresponding README.md file
- fixed reading of env variables with defaults
- gitignore emails.csv
- Import users: checking of presence of UserPrincipalName (email) instead of name
  - names are not unique and may change. Checking of UserPrincipalName
instead 
- Ignoring users who don't have @niva.no email

a lot of users like Onkel Instrument (onkel@niva.corp) in the on-prem
AD. None of the users seemed like actual users. Doing a easy filter for
now by just including niva.no emails.

A couple other users with @niva.no emails also seem like dummy or system
users, but leaving them for now
- Added unique constraint on personexternalidentifiers (personid, system)
  - I added this to the db_init script, which runs on every deploy. Is this fine? I run a alter table, but what happens if we alter table to the same state as before? This can of course fail if we have duplicates in the table, but that is a good thing I think

RmStorm commented on src/odm2_postgres_api/queries/core_queries.py:
```diff
@@ -20,6 +26,25 @@ async def insert_pydantic_object(conn: asyncpg.connection, table_name: str, pyda
+
+    if row:
+        return PersonExtended(**row)
+    else:
```
no else: needed

RmStorm commented on src/user_import/import_users.py:
```diff
@@ -0,0 +1,120 @@
+    return await insert_pydantic_object(conn, "odm2.personexternalidentifiers", external_id, PersonExternalIdentifiers)
+
+
+def name_prepare(s: str):
```
I don't quite get what this does.

haakondr replied:
hehe, perhaps not the best piece of code I've written.

It was used for two things:

- replace '' values with None
- escape singlequotes as this caused trouble with my generic query by value function. BUT, I see now that this introduced a second singlequote when inserting so this was a bug. Since I no longer query this way, I removed this alltoghether

RmStorm commented on src/user_import/import_users.py:
```diff
@@ -0,0 +1,120 @@
+            else:
+                middlename = None
+
+            people = PeopleCreate(personfirstname=name[0], personmiddlename=middlename, personlastname=user.Surname)
```
statement can be moved into `if not row:` block

RmStorm commented on src/user_import/import_users.py:
```diff
@@ -0,0 +1,120 @@
+    for user in users:
+        async with conn.transaction():
+            name = user.GivenName.strip().split(" ")
+            if len(name) > 1:
```
maybe a ternary reads a little cleaner?

RmStorm commented on src/odm2_postgres_api/utils/ODM2_for_PostgreSQL.sql:
```diff
@@ -2322,3 +2322,6 @@ create index TrackResultLocations_resultid_time_idx
 
 ALTER TABLE ODM2.Annotations ADD COLUMN annotationjson json;
+
+ALTER TABLE ODM2.personexternalidentifiers
```
we should check the other external identifier tables too?

haakondr replied:
it is not present on methodexternalidentifiers. Could make sense to do it there as well?

haakondr replied:
Alright, fixed the minor things from your feedback, squashed everything together into one commit and rewrote the commit summary. Will merge to nivatest when/if test is green

haakondr commented:
inserted users into nivatest. Script output:

2020-07-31 14:11:42,767 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="alger2@niva.corp", GivenName="", Surname="alger2", SamAccountName="alger2"]
2020-07-31 14:11:42,768 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="VALAB@niva.corp", GivenName="VALAB", Surname="", SamAccountName="VALAB"]
2020-07-31 14:11:42,769 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="EKA@niva-dk.dk", GivenName="Emilie", Surname="Kallenbach", SamAccountName="EKA"]
2020-07-31 14:11:42,769 INFO    import_users.py:109:read_users_csv, process=16625, thread=140058305929344: Ignoring user due to lacking firstname or surname [UserPrincipalName="SATPROC@niva.no", GivenName="Satproc", Surname="", SamAccountName="SATPROC"]
2020-07-31 14:11:42,770 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Test@niva.corp", GivenName="Test", Surname="Tester", SamAccountName="Test"]
2020-07-31 14:11:42,771 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Steen.Knudsen@niva-dk.dk", GivenName="Steen", Surname="Knudsen", SamAccountName="SWK"]
2020-07-31 14:11:42,771 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="_Template_Osl-OFL2@niva.corp", GivenName="_Template_Oslo-OFL2", Surname="", SamAccountName="_Template_Osl-OFL2"]
2020-07-31 14:11:42,772 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Institest@niva.corp", GivenName="Institest", Surname="", SamAccountName="Institest"]
2020-07-31 14:11:42,773 INFO    import_users.py:109:read_users_csv, process=16625, thread=140058305929344: Ignoring user due to lacking firstname or surname [UserPrincipalName="notebook-dk@niva.no", GivenName="Notebook-DK", Surname="", SamAccountName="notebook-dk"]
2020-07-31 14:11:42,774 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="FISKELABB@niva.corp", GivenName="FISKELABB", Surname="", SamAccountName="FISKELABB"]
2020-07-31 14:11:42,775 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="isoppring@niva.corp", GivenName="ISoppring", Surname="ISoppring", SamAccountName="isoppring"]
2020-07-31 14:11:42,775 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="CERAD@niva.corp", GivenName="CERAD", Surname="", SamAccountName="CERAD"]
2020-07-31 14:11:42,776 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="MPB_01@niva.corp", GivenName="Milj\u00f8pr\u00f8vebanken", Surname="", SamAccountName="MPB_01"]
2020-07-31 14:11:42,776 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="microscop311@niva.corp", GivenName="", Surname="microscop311", SamAccountName="microscop311"]
2020-07-31 14:11:42,776 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="_Template_Osl@niva.corp", GivenName="_Template_OSLO", Surname="", SamAccountName="_Template_osl"]
2020-07-31 14:11:42,777 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="MPB_02@niva.corp", GivenName="Milj\u00f8pr\u00f8vebanken", Surname="", SamAccountName="MPB_02"]
2020-07-31 14:11:42,778 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="GIS211@niva.corp", GivenName="GIS211", Surname="GIS211", SamAccountName="GIS211"]
2020-07-31 14:11:42,778 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="LABWARE@niva.corp", GivenName="Labware", Surname="Labware", SamAccountName="LABWARE"]
2020-07-31 14:11:42,778 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="IT-Test@niva.corp", GivenName="IT-Test", Surname="", SamAccountName="IT-Test"]
2020-07-31 14:11:42,778 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="BIOFOTO@niva.corp", GivenName="Biofoto", Surname="", SamAccountName="BIOFOTO"]
2020-07-31 14:11:42,779 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="FLUORE@niva.corp", GivenName="Fluore", Surname="Cytofluor", SamAccountName="FLUORE"]
2020-07-31 14:11:42,779 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ecotoxclimate@niva.corp", GivenName="ecotoxclimate", Surname="", SamAccountName="ecotoxclimate"]
2020-07-31 14:11:42,780 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="therese.harvey@niva-dk.dk", GivenName="Therese", Surname="Harvey", SamAccountName="THH"]
2020-07-31 14:11:42,780 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="JHA@niva-dk.dk", GivenName="Jesper Harbo", Surname="Andersen", SamAccountName="JHA"]
2020-07-31 14:11:42,780 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Modell@niva.corp", GivenName="Modell", Surname="", SamAccountName="Modell"]
2020-07-31 14:11:42,781 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="GC-MSD@niva.corp", GivenName="GC-MSD", Surname="Instrument", SamAccountName="GC-MSD"]
2020-07-31 14:11:42,781 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="LCMTC@niva.corp", GivenName="LCMTC", Surname="", SamAccountName="LCMTC"]
2020-07-31 14:11:42,782 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Spotlight 400 FTIR@niva.corp", GivenName="Spotlight 400 FTIR", Surname="Instrument", SamAccountName="Spotlight 400 FTIR"]
2020-07-31 14:11:42,782 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Microbeta@niva.corp", GivenName="Microbeta", Surname="Instrument", SamAccountName="Microbeta"]
2020-07-31 14:11:42,782 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Softmax@niva.corp", GivenName="Softmax", Surname="Instrument", SamAccountName="Softmax"]
2020-07-31 14:11:42,782 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ICPC@niva.corp", GivenName="ICPC", Surname="Instrument", SamAccountName="ICPC"]
2020-07-31 14:11:42,783 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Tommy@niva.corp", GivenName="Tommy", Surname="Instrument", SamAccountName="Tommy"]
2020-07-31 14:11:42,783 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="GC-TRIXIE@niva.corp", GivenName="GC-Trixie", Surname="Instrument", SamAccountName="GC-TRIXIE"]
2020-07-31 14:11:42,783 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Phoenix@niva.corp", GivenName="Phoenix", Surname="Instrument", SamAccountName="Phoenix"]
2020-07-31 14:11:42,783 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ABI-PCR@niva.corp", GivenName="ABI", Surname="LabPC", SamAccountName="ABI-PCR"]
2020-07-31 14:11:42,784 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Apollo@niva.corp", GivenName="Apollo", Surname="9000", SamAccountName="Apollo"]
2020-07-31 14:11:42,784 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="VEKT1@niva.corp", GivenName="VEKT1", Surname="", SamAccountName="VEKT1"]
2020-07-31 14:11:42,784 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="GCTOF@niva.corp", GivenName="GCTOF", Surname="Instrument", SamAccountName="GCTOF"]
2020-07-31 14:11:42,784 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ICPMS@niva.corp", GivenName="ICPMS", Surname="Instrument", SamAccountName="ICPMS"]
2020-07-31 14:11:42,784 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="QTOF@niva.corp", GivenName="QTOF", Surname="Instrument", SamAccountName="QTOF"]
2020-07-31 14:11:42,785 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Mantek@niva.corp", GivenName="Mantek", Surname="Instrument", SamAccountName="Mantek"]
2020-07-31 14:11:42,785 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Chemostat@niva.corp", GivenName="Chemostat", Surname="", SamAccountName="Chemostat"]
2020-07-31 14:11:42,785 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ICPOES@niva.corp", GivenName="ICPOES", Surname="Instrument", SamAccountName="ICPOES"]
2020-07-31 14:11:42,785 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ICPMH@niva.corp", GivenName="ICPMH", Surname="", SamAccountName="ICPMH"]
2020-07-31 14:11:42,785 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="BIO-RAD@niva.corp", GivenName="BIO-RAD", Surname="LabPC", SamAccountName="BIO-RAD"]
2020-07-31 14:11:42,786 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="JAMES@niva.corp", GivenName="James", Surname="Instrument", SamAccountName="JAMES"]
2020-07-31 14:11:42,786 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="SKALAR@niva.corp", GivenName="Skalar", Surname="Instrument", SamAccountName="SKALAR"]
2020-07-31 14:11:42,786 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Ralilal@niva.corp", GivenName="Ralilal", Surname="Instrument", SamAccountName="Ralilal"]
2020-07-31 14:11:42,786 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="HG@niva.corp", GivenName="HG", Surname="Instrument", SamAccountName="HG"]
2020-07-31 14:11:42,787 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="GC_AED@niva.corp", GivenName="GC_AED", Surname="Instrument", SamAccountName="GC_AED"]
2020-07-31 14:11:42,787 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ESTER@niva.corp", GivenName="Ester", Surname="Instrument", SamAccountName="ESTER"]
2020-07-31 14:11:42,787 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="LCMS@niva.corp", GivenName="LCMS", Surname="Instrument", SamAccountName="LCMS"]
2020-07-31 14:11:42,787 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="FIDO@niva.corp", GivenName="Fido", Surname="Instrument", SamAccountName="FIDO"]
2020-07-31 14:11:42,787 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Cary300@niva.corp", GivenName="Cary300", Surname="Instrument", SamAccountName="Cary300"]
2020-07-31 14:11:42,788 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="IC@niva.corp", GivenName="IC", Surname="Instrument", SamAccountName="IC"]
2020-07-31 14:11:42,788 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Onkel@niva.corp", GivenName="Onkel", Surname="Instrument", SamAccountName="Onkel"]
2020-07-31 14:11:42,788 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Pompel@niva.corp", GivenName="Pompel", Surname="Instrument", SamAccountName="Pompel"]
2020-07-31 14:11:42,788 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="VICTOR@niva.corp", GivenName="Victor", Surname="Instrument", SamAccountName="VICTOR"]
2020-07-31 14:11:42,788 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="FPLC@niva.corp", GivenName="FPLC", Surname="Instrument", SamAccountName="FPLC"]
2020-07-31 14:11:42,789 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="LAMBDA40P@niva.corp", GivenName="Lambda40p", Surname="Instrument", SamAccountName="LAMBDA40P"]
2020-07-31 14:11:42,789 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Fusion@niva.corp", GivenName="Fusion", Surname="Instrument", SamAccountName="Fusion"]
2020-07-31 14:11:42,789 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="MERX@niva.corp", GivenName="Merx", Surname="Instrument", SamAccountName="MERX"]
2020-07-31 14:11:42,789 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ROBOT@niva.corp", GivenName="Robot", Surname="Instrument", SamAccountName="ROBOT"]
2020-07-31 14:11:42,790 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="NYEIC@niva.corp", GivenName="NyeIC", Surname="NyeIC", SamAccountName="NYEIC"]
2020-07-31 14:11:42,790 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="SOFTMAX2@niva.corp", GivenName="Softmax2", Surname="Instrument", SamAccountName="SOFTMAX2"]
2020-07-31 14:11:42,790 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="joern.kirkegaard@niva-dk.dk", GivenName="J\u00f8rn", Surname="Kirkegaard", SamAccountName="JKI"]
2020-07-31 14:11:42,792 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="ALGER@niva.corp", GivenName="ALGER", Surname="ALGER", SamAccountName="ALGER"]
2020-07-31 14:11:42,792 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="PHOTO@niva.corp", GivenName="Photo", Surname="PC", SamAccountName="PHOTO"]
2020-07-31 14:11:42,793 INFO    import_users.py:109:read_users_csv, process=16625, thread=140058305929344: Ignoring user due to lacking firstname or surname [UserPrincipalName="nivafeltpc@niva.no", GivenName="", Surname="nivafeltpc", SamAccountName="nivafeltpc"]
2020-07-31 14:11:42,793 INFO    import_users.py:109:read_users_csv, process=16625, thread=140058305929344: Ignoring user due to lacking firstname or surname [UserPrincipalName="akvaplan@niva.no", GivenName="Akvaplan", Surname="", SamAccountName="akvaplan"]
2020-07-31 14:11:42,793 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="TOXRAT@niva.corp", GivenName="Toxrat", Surname="Toxrat", SamAccountName="TOXRAT"]
2020-07-31 14:11:42,793 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Kurs1@niva.corp", GivenName="Kurs1", Surname="Kursbruker", SamAccountName="Kurs1"]
2020-07-31 14:11:42,794 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="mathias.brink.kjeldgaard@niva-dk.dk", GivenName="Mathias", Surname="Brink Kjeldgaard", SamAccountName="MBK"]
2020-07-31 14:11:42,795 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="UTSTYR@niva.corp", GivenName="Utstyr", Surname="Sentralen", SamAccountName="UTSTYR"]
2020-07-31 14:11:42,796 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="LAB-OMF@niva.corp", GivenName="LAB-OMF", Surname="Pr\u00f8vemottak", SamAccountName="LAB-OMF"]
2020-07-31 14:11:42,796 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="Instrument@niva.corp", GivenName="Instrument", Surname="PC", SamAccountName="Instrument"]
2020-07-31 14:11:42,796 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="CJM@niva-dk.dk", GivenName="Ciaran Joseph", Surname="Murray", SamAccountName="CJM"]
2020-07-31 14:11:42,797 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="testa@niva.corp", GivenName="testa", Surname="Hanssen", SamAccountName="testa"]
2020-07-31 14:11:42,797 INFO    import_users.py:106:read_users_csv, process=16625, thread=140058305929344: Ignoring user without @niva.no email [UserPrincipalName="MERXTOTAL@niva.corp", GivenName="MERXTOTAL", Surname="", SamAccountName="MERXTOTAL"]

haakondr commented:
I see that I did indeed create duplicates for already existing "people":

select p.*, a.primaryemail from odm2.people p inner join odm2.affiliations a on p.personid=a.personid where p.personfirstname='Håkon'

| personid | personfirstname | personmiddlename | personlastname | primaryemail |
| :--- | :--- | :--- | :--- | :--- |
| 3 | Håkon | Drolsum | Røkenes | hakon.rokenes@niva.no |
| 31 | Håkon | Drolsum | Røkenes | Hakon.Drolsum.Rokenes@niva.no |

I think I might be fetching the wrong email in niva port.

haakondr commented:
seems like I am the only victim so far
```
select personfirstname, count(*) from odm2.people group by personfirstname, personlastname having count(*) > 1
```
| personfirstname | c |
| :--- | :--- |
| Håkon | 2 |
